From 91891162c89ee1cc6df878c9a5ddc37bc2051fb3 Mon Sep 17 00:00:00 2001
From: Michael Glembotzki <Michael.Glembotzki@iris-sensing.com>
Date: Fri, 6 Dec 2024 10:10:53 +0000
Subject: [PATCH] nvmem: imx-ocotp-fsb-s400: Add property
 fsb_disable_shadowed_read

Add property fsb_disable_shadowed_read to disable fuse shadow register
read and enforce fuse sensing.

Upstream-Status: Inappropriate
Signed-off-by: Michael Glembotzki <Michael.Glembotzki@iris-sensing.com>
---
 arch/arm64/boot/dts/freescale/imx93-matrixup.dts | 5 +++++
 drivers/nvmem/imx-ocotp-fsb-s400.c               | 2 +-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/freescale/imx93-matrixup.dts b/arch/arm64/boot/dts/freescale/imx93-matrixup.dts
index 44bc63b87fd73..7bcbdb93eac09 100644
--- a/arch/arm64/boot/dts/freescale/imx93-matrixup.dts
+++ b/arch/arm64/boot/dts/freescale/imx93-matrixup.dts
@@ -593,3 +593,8 @@ cap_device {
 		status = "okay";
 	};
 };
+
+&aonmix_ns_gpr {
+	/* disable shadow register read to enforce fuse sensing */
+	fsb_disable_shadowed_read;
+};
diff --git a/drivers/nvmem/imx-ocotp-fsb-s400.c b/drivers/nvmem/imx-ocotp-fsb-s400.c
index 7fd3990f00ea9..698c0c62dca77 100644
--- a/drivers/nvmem/imx-ocotp-fsb-s400.c
+++ b/drivers/nvmem/imx-ocotp-fsb-s400.c
@@ -372,7 +372,7 @@ static int imx_fsb_s400_fuse_probe(struct platform_device *pdev)
 			return -ENOMEM;
 
 		v = readl_relaxed(reg + FUSE_ACC_DIS);
-		if (v & BIT(0))
+		if (v & BIT(0) || of_property_read_bool(np, "fsb_disable_shadowed_read"))
 			fuse->fsb_read_dis = true;
 		else
 			fuse->fsb_read_dis = false;
-- 
2.47.1

